File: ios/peeriomobile/RNSodium.m
//
//  RNSodium
//  RNSodium
//
//  Created by Samvel Avanesov on 05/01/2017.
//  Copyright © 2017 Technologies Peerio Inc. All rights reserved.
//
#include <sys/types.h>
#include <sys/sysctl.h>
#include <React/RCTLog.h>
#import "RNSodium.h"

#include "libsodium-ios/include/sodium.h"

@implementation RNSodium

RCT_EXPORT_MODULE();

+ (void)initialize {
    int result = sodium_init();
    NSLog(@"RNSodium.m: sodium init %d", result);
}

RCT_REMAP_METHOD(scrypt,
    passwordB64:(NSString*)passwordB64
    saltB64:(NSString*)saltB64
    p_N: (nonnull NSNumber* )p_N
    p_r: (nonnull NSNumber* )p_r
    p_p: (nonnull NSNumber* )p_p
    p_dkLen: (nonnull NSNumber*)p_dkLen
    resolver: (RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
{
    NSData *password = [[NSData alloc] initWithBase64EncodedString:passwordB64 options:0];
    NSData *salt = [[NSData alloc] initWithBase64EncodedString:saltB64 options:0];
    unsigned long N = [p_N unsignedIntegerValue];
    unsigned int r = [p_r unsignedIntValue];
    unsigned int p = [p_p unsignedIntValue];
    unsigned int dkLen = [p_dkLen unsignedIntValue];
    NSMutableData *buffer = [NSMutableData dataWithLength:dkLen];
    int result = crypto_pwhash_scryptsalsa208sha256_ll(
      password.bytes, password.length,
      salt.bytes, salt.length,
      N, r, p,
      buffer.mutableBytes, buffer.length
    );
    if (result == 0) {
        NSString* resultB64 = [buffer base64Encoding];
        resolve(resultB64);
    } else {
        reject(@"RNSodium.m: scrypt error", nil, nil);
    }
}

RCT_REMAP_METHOD(signDetached,
    messageB64:(NSString*)messageB64
    secretKeyB64:(NSString*)secretKeyB64
    resolver: (RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
{
    NSData *message = [[NSData alloc] initWithBase64EncodedString:messageB64 options:0];
    NSData *secretKey = [[NSData alloc] initWithBase64EncodedString:secretKeyB64 options:0];
    unsigned long long signLength = 64;
    NSMutableData *buffer = [NSMutableData dataWithLength:signLength];
    @try {
      int result = crypto_sign_detached(
        buffer.mutableBytes,
        &signLength,
        message.bytes,
        message.length,
        secretKey.bytes
      );
      if (result == 0) {
          NSString* resultB64 = [buffer base64Encoding];
          resolve(resultB64);
      } else {
          reject(@"RNSodium.m: signDetached error", nil, nil);
      }
    } @catch (NSException*) {
      reject(@"RNSodium.m: signDetached exception", nil, nil);
    }
}

RCT_REMAP_METHOD(verifyDetached,
    messageB64:(NSString*)messageB64
    signatureB64:(NSString*)signatureB64
    publicKeyB64:(NSString*)publicKeyB64
    resolver: (RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
{
    @try {
    NSData *message = [[NSData alloc] initWithBase64EncodedString:messageB64 options:0];
    NSData *signature = [[NSData alloc] initWithBase64EncodedString:signatureB64 options:0];
    NSData *publicKey = [[NSData alloc] initWithBase64EncodedString:publicKeyB64 options:0];
    int signLength = 64;
      int result = crypto_sign_verify_detached(
        signature.bytes,
        message.bytes,
        message.length,
        publicKey.bytes
      );
      if (result == 0) {
          resolve([NSNumber numberWithInt:1]);
      } else {
          reject(@"RNSodium.m: verifyDetached error", nil, nil);
      }
    } @catch (NSException*) {
        reject(@"RNSodium.m: verifyDetached exception", nil, nil);
    }
}

@end

File: android/app/src/main/java/com/peerio/messenger/RNSodium.java
package com.peerio.messenger;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.Promise;

import android.util.Base64;

import org.libsodium.jni.Sodium;

import java.io.Console;

import static org.libsodium.jni.NaCl.sodium;

/* import static org.libsodium.jni.SodiumJNI.crypto_pwhash_scryptsalsa208sha256_ll;
import static org.libsodium.jni.SodiumJNI.sodium_init; */

public class RNSodium extends ReactContextBaseJavaModule {

    public RNSodium(final ReactApplicationContext reactContext) {
        super(reactContext);
        this._sodium = sodium();
    }

    Sodium _sodium = null;

    @Override
    public String getName() {
        return "RNSodium";
    }

    @ReactMethod
    public void scrypt(
            final String passwordB64,
            final String saltB64,
            final int N,
            final int r,
            final int p,
            final int dkLen,
            final Promise promise
    ) {
        byte[] password = Base64.decode(passwordB64, Base64.DEFAULT);
        byte[] salt = Base64.decode(saltB64, Base64.DEFAULT);
        byte[] buffer = new byte[dkLen];
        int result = _sodium.crypto_pwhash_scryptsalsa208sha256_ll(
                password,
                password.length,
                salt,
                salt.length,
                N,
                r,
                p,
                buffer,
                buffer.length
        );

        if (result == 0) {
            String resultB64 = Base64.encodeToString(buffer, Base64.DEFAULT);
            promise.resolve(resultB64);
        } else {
            promise.reject("RNSodium.java: scrypt error", "RNSodium.java: scrypt error");
        }
    }

    @ReactMethod
    public void signDetached(
            final String messageB64,
            final String secretKeyB64,
            final Promise promise
    ) {
        byte[] message = Base64.decode(messageB64, Base64.DEFAULT);
        byte[] secretKey = Base64.decode(secretKeyB64, Base64.DEFAULT);
        final int signLength = 64;
        byte[] buffer = new byte[signLength];
        int[] bufferLength = new int[1];
        bufferLength[0] = signLength;
        int result = _sodium.crypto_sign_detached(
                buffer,
                bufferLength,
                message,
                message.length,
                secretKey
        );

        if (result == 0) {
            String resultB64 = Base64.encodeToString(buffer, Base64.DEFAULT);
            promise.resolve(resultB64);
        } else {
            promise.reject("RNSodium.java: sign_detached error", "RNSodium.java: sign_detached error");
        }
    }

    @ReactMethod
    public void verifyDetached(
            final String messageB64,
            final String signatureB64,
            final String publicKeyB64,
            final Promise promise
    ) {
        byte[] message = Base64.decode(messageB64, Base64.DEFAULT);
        byte[] signature = Base64.decode(signatureB64, Base64.DEFAULT);
        byte[] publicKey = Base64.decode(publicKeyB64, Base64.DEFAULT);
        int result = _sodium.crypto_sign_verify_detached(
                signature,
                message,
                message.length,
                publicKey
        );

        if (result == 0) {
            promise.resolve(true);
        } else {
            promise.reject("RNSodium.java: verify_detached error", "RNSodium.java: verify_detached error");
        }
    }
}
